sql/analyzer/rules.go

```diff
@@ -411,3 +412,71 @@ func pushdown(a *Analyzer, n sql.Node) (sql.Node, error) {
 		return node, nil
 	})
 }
+
+func castCompares(a *Analyzer, n sql.Node) (sql.Node, error) {
```

Comment:
What do you think about putting this logic inside the `Comparison` `Compare` method instead?

`Compare` has access to the types and it could handle them itself using the `Convert` method of the types. The less rules we need, the better.

Reply:
The rule that I'm missing here is one to cast literals in comparisons when the types on left and right are not equals.

Reply:
`Literal`s are like any other `Expression` and they have certain type. The rule can't be cast always the literal to the other type, at least mysql doesn't work in this way.

For example, when the types at both side of the comparison are different, mysql cast to numbers if some of the type of them is a number, independently if there is a literal or not.

As you can see in this example, mysql cast the text values from the `num` column to numbers, (the casting of the text values evals to 0 because they don't represent numbers):
```
MySQL [testing]> select * from tabletest;
+----+------+------+
| id | name | num  |
+----+------+------+
|  1 | a    |    1 |
|  2 | b    |    2 |
|  3 | c    |    3 |
|  4 | d    |    4 |
+----+------+------+
4 rows in set (0.00 sec)

MySQL [testing]> select * from tabletest where name > 0;
Empty set (0.00 sec)

MySQL [testing]> select * from tabletest where name < 0;
Empty set (0.01 sec)

MySQL [testing]> select * from tabletest where name = 0;
+----+------+------+
| id | name | num  |
+----+------+------+
|  1 | a    |    1 |
|  2 | b    |    2 |
|  3 | c    |    3 |
|  4 | d    |    4 |
+----+------+------+
4 rows in set (0.00 sec)

```